### PR TITLE
fix: pagination over time + tooltip

### PR DESCRIPTION
**Pull request by jmbuss**
## Overview
Allow the user to paginate across time. Pagination will allow users to shift the time displayed on the dashboard widgets by that given range. For example, if the user is viewing data from the last week, they can paginate forward to see next weeks data or backwards to see last weeks data being displayed on their dashboard widgets.

## Verifying Changes

https://github.com/awslabs/iot-app-kit/assets/145582655/cf4211c0-7c15-44be-831c-33680d24c8e0

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

**Review by tracy-french — APPROVED** [review]
Thank you for working through my feedback and making a fantastic revision! Well done!

File: packages/react-components/src/components/time-sync/viewportAdapter.tsx [hunk 1]
```diff
@@ -119,3 +119,31 @@ export const viewportToDateRange = (viewport?: Viewport): DateRangePickerProps.V
 
   return null;
 };
+
+export const getNewViewportStartDate = (value: DateRangePickerProps.RelativeValue): Date => {
```
Comment (chejimmy) [inline packages/react-components/src/components/time-sync/viewportAdapter.tsx:123]: nit: the name doesn't capture how the function get the new start date, maybe like `getBackwardRelativeStartDate`? 🤔

File: packages/react-components/src/components/time-sync/timeSelection.tsx [hunk 2]
```diff
@@ -52,31 +56,97 @@ const messages: ViewportMessages = {
...
+    }
+  };
 
-    setViewport(dateRangeToViewport(value), 'date-picker');
+  const handlePaginateBackward = () => {
+    const value = viewportToDateRange(viewport);
+    if (!value) return;
+    if (value.type == 'absolute') {
+      const duration = new Date(value.endDate).getTime() - new Date(value.startDate).getTime();
+      const newStart = new Date(new Date(value.startDate).getTime() - duration);
```
Comment (chejimmy) [inline packages/react-components/src/components/time-sync/timeSelection.tsx:93]: edge case: does it handle it gracefully when the `startDate` has a value of the oldest available date?

**Review by chejimmy — COMMENTED** [review]
Generally looks good to me! 👍
Just have a question about the edge case

File: packages/react-components/src/components/time-sync/timeSelection.tsx (hunk 2 above)
Comment (ssjagad) [reply packages/react-components/src/components/time-sync/timeSelection.tsx:93]: Thought about that. UTC starts at Jan 1st, 1970 and the UTC date object can go all the way back to 1900. Might be out of reach by customer pagination.

**Comment by ssjagad** [comment]
Since, we changed the x-axis to go into the future, the disable feature of the forward pagination has been removed in the latest commit

**Review by corteggiano — APPROVED** [review]

File: packages/react-components/src/components/chart/hooks/useDataZoom.ts [hunk 3]
```diff
@@ -32,13 +32,20 @@ export const useDataZoom = (chartRef: MutableRefObject<EChartsType | null>, view
     viewportInMsRef.current = viewportInMs;
   }, [viewportInMs]);
 
+  // handle live mode
   useEffect(() => {
     const chart = chartRef.current;
-    if (chart && (!isScrolling || lastUpdatedBy === 'date-picker')) {
+    if (chart && !isScrolling && viewportInMs.isDurationViewport) {
```
[the remark this answers is not in the document]
Comment (mnischay) [reply packages/react-components/src/components/chart/hooks/useDataZoom.ts:38]: is this the only diff ? setIsScrolling(false);

File: packages/react-components/src/components/time-sync/timeSelection.tsx [hunk 4]
```diff
@@ -52,31 +60,83 @@ const messages: ViewportMessages = {
...
+          relativeOptions={relativeOptions}
+          isValidRange={rangeValidator({ dateRangeIncompleteError, dateRangeInvalidError })}
+          i18nStrings={i18nStrings}
+          placeholder={placeholder}
+        />
+        {isPaginationEnabled && (
+          <div className='tooltip' aria-label='paginate-foward'>
+            <Button iconName='caret-right-filled' onClick={handlePaginateForward} />
+
+            <Tooltip
```
Comment (mnischay) [inline packages/react-components/src/components/time-sync/timeSelection.tsx:134]: same as above

File: packages/react-components/src/components/time-sync/viewportAdapter.tsx [hunk 5]
```diff
@@ -119,3 +119,31 @@ export const viewportToDateRange = (viewport?: Viewport): DateRangePickerProps.V
 
   return null;
 };
+
+export const getViewportStartOnBackwardRelative = (value: DateRangePickerProps.RelativeValue): Date => {
```
Comment (mnischay) [inline packages/react-components/src/components/time-sync/viewportAdapter.tsx:123]: can you please add some tests for this ?

File: packages/react-components/src/components/tooltip/tooltip.css [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  /* margin-left: -60px;  */
+  visibility: hidden;
+  text-align: center;
+  padding: 5px 0;
+  border-radius: 6px;
+}
+
+/* Can't style pseudo selectors with inline styling */
+.tooltip .tooltiptext::before,
+.tooltip .tooltiptext::after {
```
Comment (mnischay) [inline packages/react-components/src/components/tooltip/tooltip.css:24]: optional: for before and after worth using the event handlers and design tokens ?

File: packages/react-components/src/components/tooltip/tooltip.css [hunk 7]
```diff
@@ -0,0 +1,51 @@
+.tooltip {
```
Comment (mnischay) [inline packages/react-components/src/components/tooltip/tooltip.css:1]: this is used in multiple placess, especially the resource explorer, make sure this hasnt broken those

File: packages/react-components/src/components/tooltip/tooltip.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
```
Comment (mnischay) [inline packages/react-components/src/components/tooltip/tooltip.tsx:1]: we already have a tooltip i think ? can you please check ?

**Review by mnischay — CHANGES_REQUESTED** [review]

File: packages/react-components/src/components/tooltip/tooltip.css [hunk 9]
```diff
@@ -21,7 +21,7 @@
 }
 
 .bottom {
-  top: 90%;
+  top: 120%;
```
[the remark this answers is not in the document]
Comment (ssjagad) [reply packages/react-components/src/components/tooltip/tooltip.css:24]: it does not affect any other tooltip. This is the only tooltip that uses .bottom class

File: packages/react-components/src/components/chart/hooks/useDataZoom.ts [hunk 10]
```diff
@@ -32,10 +32,14 @@ export const useDataZoom = (chartRef: MutableRefObject<EChartsType | null>, view
     viewportInMsRef.current = viewportInMs;
   }, [viewportInMs]);
 
+  // handle live mode + pagination
   useEffect(() => {
     const chart = chartRef.current;
-    if (chart && (!isScrolling || lastUpdatedBy === 'date-picker')) {
-      setIsScrolling(false);
+    if (chart && !isScrolling) {
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/components/chart/hooks/useDataZoom.ts:38]: yeah, we went through everything online 👍

**Review by mnischay — APPROVED** [review]

**Pull request merged**